client/src/controllers/ProgressController.ts

```diff
@@ -0,0 +1,138 @@
+ *   data-action="w-progress#activate"
+ * >
+ *  {% icon name="spinner" %}
+ *  <em data-w-progress-target="label">{% trans 'Sign in' %}</em>
+ * </button>
+ */
+export class ProgressController extends Controller {
+  static classes = ['active'];
+  static targets = ['label'];
+  static values = {
+    active: { default: gettext('Loading'), type: String },
+    durationSeconds: { default: 30, type: Number },
+    label: { default: '', type: String },
+    loading: { default: false, type: Boolean },
+  };
+
+  declare activeClass: string;
+  /** Label to use when loading */
+  declare activeValue: string;
+  declare durationSecondsValue: number;
+  /** Label to store the original text on the button */
+  declare labelValue: string;
+  declare loadingValue: boolean;
+  declare readonly hasActiveClass: boolean;
+  declare readonly hasLabelTarget: boolean;
+  declare readonly labelTarget: HTMLElement;
+  timer?: number;
+
+  static afterLoad(identifier: string, application: Application) {
+    const { controllerAttribute } = application.schema;
+    const { actionAttribute } = application.schema;
+
+    document.addEventListener(
+      'DOMContentLoaded',
+      () => {
+        document
+          .querySelectorAll(
+            `.button-longrunning:not([${controllerAttribute}~='${identifier}'])`,
+          )
+          .forEach((button) => {
+            button.setAttribute(controllerAttribute, identifier);
+            button.setAttribute(actionAttribute, `${identifier}#activate`);
+
+            const activeText = button.getAttribute('data-clicked-text');
+            if (activeText) {
+              button.setAttribute(
+                `data-${identifier}-active-value`,
+                activeText,
+              );
+              button.removeAttribute('data-clicked-text');
+            }
+
+            const labelElement = button.querySelector('em');
+            if (labelElement) {
+              labelElement.setAttribute(`data-${identifier}-target`, 'label');
+            }
+
+            button.setAttribute(
+              `data-${identifier}-duration-seconds-value`,
+              '30',
+            );
+          });
+      },
+      { once: true, passive: true },
+    );
+  }
+
+  activate() {
+    // If client-side validation is active on this form, and is going to block submission of the
+    // form, don't activate the spinner
+    const form = this.element.closest('form');
+
+    if (
+      form &&
+      form.checkValidity &&
+      !form.noValidate &&
+      !form.checkValidity()
+    ) {
+      return;
+    }
+
+    window.setTimeout(() => {
+      this.loadingValue = true;
+
+      const durationMs = this.durationSecondsValue * 1000;
+
+      this.timer = window.setTimeout(() => {
+        this.loadingValue = false;
+      }, durationMs);
+    });
+  }
+
+  loadingValueChanged(isLoading: boolean) {
+    const activeClass = this.hasActiveClass
+      ? this.activeClass
+      : 'button-longrunning-active';
```

Comment:
```suggestion
      : `${DEFAULT_CLASS}-active';
```

Let's use a constant for the `DEFAULT_CLASS` here and above